### PR TITLE
[WIP] Attnet subscription revamp

### PR DESCRIPTION
**Pull request by AgeManning**
This is a developer PR.

The functionality this PR enables is designed for developers only. We want to test networking with a lower number of long-lived subscriptions. This will add a hidden developer CLI flag to run on various testnets to check the viability of this model.

### CODE REVIEW

File: beacon_node/http_api/src/lib.rs [hunk 1]
```diff
@@ -2350,6 +2350,8 @@ pub fn serve<T: BeaconChainTypes>(
             },
         );
 
+    // TODO: is this needed?
+    // Think not?
```
Comment (AgeManning) [inline beacon_node/http_api/src/lib.rs]: There are currently two use cases for this:
1. The beacon chain `monitor` functionality. This allows users to monitor their attached validators. The beacon chain provides useful logs about how the attached validators are performing. 
2. For long-lived subnets in the network. 

For the context of this PR we should leave this. Because we want it to maintain the default functionality and only optionally (via hidden CLI) allow the reduced long-lived subscriptions. 

If this new option is viable on the network, i think we can have a chat with other teams about removing this API all together and lighthouse can come up with another way to monitor validators.

tldr; lets leave it for now (also less work :).

File: consensus/types/src/chain_spec.rs [hunk 2]
```diff
@@ -157,8 +157,9 @@ pub struct ChainSpec {
     pub maximum_gossip_clock_disparity_millis: u64,
     pub target_aggregators_per_committee: u64,
     pub attestation_subnet_count: u64,
-    pub random_subnets_per_validator: u64,
-    pub epochs_per_random_subnet_subscription: u64,
+    pub subnets_per_node: u8,
+    pub epochs_per_subnet_subscription: u64,
+    pub attestation_subnet_extra_bits: u8,
```
Comment (AgeManning) [inline consensus/types/src/chain_spec.rs]: maybe we still want to keep the old ones, so the default behaviour is the old behaviour

File: consensus/types/src/chain_spec.rs [hunk 3]
```diff
@@ -561,10 +575,11 @@ impl ChainSpec {
             network_id: 1, // mainnet network id
             attestation_propagation_slot_range: 32,
             attestation_subnet_count: 64,
-            random_subnets_per_validator: 1,
+            subnets_per_node: 1,
             maximum_gossip_clock_disparity_millis: 500,
             target_aggregators_per_committee: 16,
-            epochs_per_random_subnet_subscription: 256,
+            epochs_per_subnet_subscription: 256,
+            attestation_subnet_extra_bits: 6,
```
Comment (AgeManning) [inline consensus/types/src/chain_spec.rs:608]: tempted to start with 0, and we can increase to see how we go. 

Maybe we should start with 6 and if it doesn't work lower it. We will prob have to increase number of parallel discovery queries

File: consensus/types/src/subnet_id.rs [hunk 4]
```diff
@@ -71,6 +72,32 @@ impl SubnetId {
...
             .into())
     }
+
+    pub fn compute_subnets_for_epoch<'a, T: EthSpec>(
+        node_id: ethereum_types::U256,
+        epoch: Epoch,
+        spec: &'a ChainSpec,
+    ) -> Result<impl Iterator<Item = SubnetId> + 'a, &'static str> {
+        let node_id_prefix =
+            (node_id >> (256 - spec.attestation_subnet_extra_bits as usize)).as_usize();
```
Comment (AgeManning) [inline consensus/types/src/subnet_id.rs]: Should this be:
```rust
(256 - spec.attestation_subnet_prefix_bits() as usize)
```
?

File: beacon_node/network/src/subnet_service/attestation_subnets.rs [hunk 5]
```diff
@@ -25,13 +23,12 @@ use crate::metrics;
...
 const MIN_PEER_DISCOVERY_SLOT_LOOK_AHEAD: u64 = 2;
-/// The time (in slots) before a last seen validator is considered absent and we unsubscribe from the random
-/// gossip topics that we subscribed to due to the validator connection.
-const LAST_SEEN_VALIDATOR_TIMEOUT: u32 = 150;
 /// The fraction of a slot that we subscribe to a subnet before the required slot.
 ///
 /// Note: The time is calculated as `time = seconds_per_slot / ADVANCE_SUBSCRIPTION_TIME`.
-const ADVANCE_SUBSCRIBE_TIME: u32 = 3;
+/// TODO: apply too?
+// const ADVANCE_SUBSCRIBE_TIME: u32 = 3;
```
Comment (AgeManning) [inline beacon_node/network/src/subnet_service/attestation_subnets.rs]: I thought we used this for regular (no long-lived) subnet subscriptions? Perhaps not?

File: beacon_node/network/src/subnet_service/attestation_subnets.rs [hunk 6]
```diff
@@ -49,12 +46,18 @@ pub struct AttestationService<T: BeaconChainTypes> {
...
     events: VecDeque<SubnetServiceMessage>,
 
+    /// Our Discv5 node_id.
+    node_id: ethereum_types::U256,
+
+    /// Long lived subnets for which we are currently subscribed.
+    long_lived_subnets: HashSet<SubnetId>,
+
+    /// Future used to manage subscribing and unsubscribing from long lived subnets.
+    next_epoch: Pin<Box<tokio::time::Sleep>>,
```
Comment (AgeManning) [inline beacon_node/network/src/subnet_service/attestation_subnets.rs]: maybe we could be smarter and only do the calculation every `SUBNET_DURATION_IN_EPOCHS`. ?

File: beacon_node/network/src/subnet_service/attestation_subnets.rs [hunk 7]
```diff
@@ -92,42 +89,101 @@ impl<T: BeaconChainTypes> AttestationService<T> {
...
+            SubnetServiceMessage::Unsubscribe(Subnet::Attestation(*subnet))
+        });
+
+        let subnets_to_subscribe = subnets.difference(&self.long_lived_subnets).map(|subnet| {
+            debug!(self.log, "Subscribing to long lived subnet"; "subnet" => ?subnet);
+            SubnetServiceMessage::Subscribe(Subnet::Attestation(*subnet))
+        });
+
+        self.events.extend(subnets_to_unsubscribe);
+        self.events.extend(subnets_to_subscribe);
```
Comment (AgeManning) [inline beacon_node/network/src/subnet_service/attestation_subnets.rs]: for backwards compatibilty we probably should send the `EndAdd` and `EnrRemove` events

**Review by AgeManning — COMMENTED** [review]
This looks pretty good to me. 

The one thing is that we would need to maintain backwards compatibility by default and optionally allow us to switch to this mode for dev testing. 

Probably means we need to keep some of the old code that is removed.

File: consensus/types/src/subnet_id.rs (hunk 4 above)
Comment (divagant-martian) [reply consensus/types/src/subnet_id.rs]: yep, ty sir

File: beacon_node/network/src/subnet_service/attestation_subnets.rs (hunk 5 above)
Comment (divagant-martian) [reply beacon_node/network/src/subnet_service/attestation_subnets.rs]: well it's flagged as unused code if I remove the rest so..

File: beacon_node/network/src/subnet_service/attestation_subnets.rs (hunk 6 above)
Comment (divagant-martian) [reply beacon_node/network/src/subnet_service/attestation_subnets.rs]: good idea

File: beacon_node/network/src/subnet_service/attestation_subnets.rs (hunk 7 above)
Comment (divagant-martian) [reply beacon_node/network/src/subnet_service/attestation_subnets.rs]: I think we need to do this anyway

**Comment by divagant-martian** [comment]
after better understanding how the attestation service works I'm closing this to take a different approach
